mobile-widgets/qmlprefs.cpp

```diff
@@ -0,0 +1,197 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "qmlprefs.h"
+#include "qmlmanager.h"
+
+#include "core/membuffer.h"
+#include "core/subsurface-qt/SettingsObjectWrapper.h"
+#include "core/gpslocation.h"
+
+
+/*** Global and constructors ***/
+QMLPrefs *QMLPrefs::m_instance = NULL;
+
+QMLPrefs::QMLPrefs() :
+	m_cloudPassword(""),
+	m_cloudPin(""),
+	m_cloudUserName(""),
+	m_credentialStatus(CS_UNKNOWN),
+	m_developer(false),
+	m_distanceThreshold(1000),
+	m_oldStatus(CS_UNKNOWN),
+	m_showPin(false),
+	m_timeThreshold(60)
+{
+	if (!m_instance)
+		m_instance = this;
+}
+
+QMLPrefs::~QMLPrefs()
+{
+	m_instance = NULL;
+}
+
+QMLPrefs *QMLPrefs::instance()
+{
+	return m_instance;
+}
+
+
+/*** public functions ***/
+const QString QMLPrefs::cloudPassword() const
+{
+	return m_cloudPassword;
+}
+
+void QMLPrefs::setCloudPassword(const QString &cloudPassword)
+{
+	m_cloudPassword = cloudPassword;
+	emit cloudPasswordChanged();
+}
+
+const QString QMLPrefs::cloudPin() const
+{
+	return m_cloudPin;
+}
+
+void QMLPrefs::setCloudPin(const QString &cloudPin)
+{
+	m_cloudPin = cloudPin;
+	emit cloudPinChanged();
+}
+
+const QString QMLPrefs::cloudUserName() const
+{
+	return m_cloudUserName;
+}
+
+void QMLPrefs::setCloudUserName(const QString &cloudUserName)
+{
+	m_cloudUserName = cloudUserName.toLower();
+	emit cloudUserNameChanged();
+}
+
+QMLPrefs::cloud_status_qml QMLPrefs::credentialStatus() const
+{
+	return m_credentialStatus;
+}
+
+void QMLPrefs::setCredentialStatus(const cloud_status_qml value)
+{
+	if (m_credentialStatus != value) {
+		setOldStatus(m_credentialStatus);
+		if (value == CS_NOCLOUD) {
+			QMLManager::instance()->appendTextToLog("Switching to no cloud mode");
+			set_filename(NOCLOUD_LOCALSTORAGE);
+			clearCredentials();
+		}
+		m_credentialStatus = value;
+		emit credentialStatusChanged();
+	}
+}
+
+void QMLPrefs::setDeveloper(bool value)
+{
+	m_developer = value;
+	emit developerChanged();
+}
+
+int QMLPrefs::distanceThreshold() const
```

Comment:
All these getter-functions look so horribly pointless. Are we sure that they are needed?

I vaguely remember that the Q_PROPERTY stuff does not need explicit getters.

Reply:
Yes Q_PROPERTY can be written with MEMBER, as you can see in several places, but I needed to add getter functions for those variables used by other objects (in this case QMLManager).

I preferred a few getter functions to make QMLManager friend, but have no problem changing to that.

Medium term, all these variables should go into prefs and thus be shared without getter.